### PR TITLE
ENH: RESTful client supports custom model APIs

### LINKED ISSUE #355: ENH: RESTful client supports custom model APIs
### Is your feature request related to a problem? Please describe
Including `register_model`, `unregister_model`, and `list_model_registrations`.

### PR DESCRIPTION
**Pull request by jiayini1119**
Resolves #355.

### CODE REVIEW

File: xinference/client.py [hunk 1]
```diff
@@ -763,3 +763,38 @@ def get_model(self, model_uid: str) -> RESTfulModelHandle:
             return RESTfulChatModelHandle(model_uid, self.base_url)
         else:
             raise ValueError(f"Unrecognized model ability: {desc['model_ability']}")
+
+    def register_model(self, model_type: str, model: str, persist: bool):
+        url = f"{self.base_url}/v2/models/register_model"
```
Comment (UranusSeven) [inline xinference/client.py:768]: Could you also add docstrings to custom model APIs?

File: xinference/core/restful_api.py [hunk 2]
```diff
@@ -266,6 +277,20 @@ def serve(self):
             methods=["POST"],
             response_model=ChatCompletion,
         )
+
+        # for custom models
+        self._router.add_api_route(
+            "/v2/models/register_model", self.register_model, methods=["POST"]
```
Comment (UranusSeven) [inline xinference/core/restful_api.py:283]: Adding new functionality won't break backward compatiblity. Thus, it is safe to use `v1` for custom model APIs.

Comment (UranusSeven) [inline xinference/core/restful_api.py:283]: Consider using `/v1/model_registrations` for better adherence to RESTful API design. Here are the suggested endpoints:

Register an LLM: POST /v1/model_registrations/llm
Unregister an LLM: DELETE /v1/model_registrations/llm
List LLM registrations: GET /v1/model_registrations/llm
Get an LLM registration: GET /v1/model_registrations/llm/${model_name}

Comment (jiayini1119) [reply xinference/core/restful_api.py:283]: udpated in the latest commit

Comment (jiayini1119) [reply xinference/core/restful_api.py:283]: udpated in the latest commit

**Review by UranusSeven — APPROVED** [review]

**Review by UranusSeven — APPROVED** [review]

**Pull request merged**